### PR TITLE
Support linear handles in Oak Runtime.

### PR DESCRIPTION
**Pull request by wildarch**
This also migrates the hello_world example to use linear-handles.

Note: I was planning to migrate all of the examples first, but that is causing tests to fail, as some examples have a dev dependency on `oak_runtime`.

Tracking issue: #1854.

# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by
        [Cloudbuild](/cloudbuild.yaml)
  - [ ] I have updated [documentation](/docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: oak_runtime/src/lib.rs [hunk 1]
```diff
@@ -394,10 +394,14 @@ impl AuxServer {
     ) -> Self {
         let (termination_notificiation_sender, termination_notificiation_receiver) =
             tokio::sync::oneshot::channel::<()>();
+        let runtime_proxy = runtime.clone().proxy_for_new_node(name);
```
[the remark this answers is not in the document]
Comment (wildarch) [reply oak_runtime/src/lib.rs:397]: The AuxServer is definitely on its own thread, and it communicates with other nodes using handles, so I would assume it qualifies as a node?

The reason I am creating a proxy here is that the thread local variable stores a proxy, not an `Arc<Runtime>`

File: oak_runtime/src/proxy.rs [hunk 2]
```diff
@@ -624,4 +625,62 @@ impl RuntimeProxy {
...
+    ///
+    /// In particular, make sure to call this when a new node is created.
+    pub fn set_as_current(&self) {
+        RUNTIME_PROXY.with(|proxy| {
+            proxy.replace(Some(self.clone()));
+        })
+    }
+}
+
+std::thread_local! {
```
[the remark this answers is not in the document]
Comment (wildarch) [reply oak_runtime/src/proxy.rs:655]: You're right that we require a call to `set_as_current` in every thread, but I am not sure if there is a way to make this cleaner. Threads are created in different ways, sometimes we need to call it in tokio worker threads, sometimes in a plain thread, so I don't think we can create one function that fits all use cases (and we would still have to get everyone starting a new thread to use it).

File: oak_runtime/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rbehjati) [reply oak_runtime/src/lib.rs:397]: Are you sure it communicates with other nodes using handles? I think it should not. Also, it does not implement `Node`. 
Do any of the tests fail if you remove `proxy_for_new_node(name)`?

File: oak_runtime/src/proxy.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rbehjati) [reply oak_runtime/src/proxy.rs:655]: Yeah. Unfortunately that is what I thought too, but I was hoping you might have a solution :)

File: oak_runtime/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rbehjati) [reply oak_runtime/src/lib.rs:397]: This might also explain why `node 1` does not exist anymore (in `abitest`). We have two AuxServers, one for metrics and one for introspection. I think these servers are being created before any other node. I am guessing with the call to `proxy_for_new_node` here, these servers consume the first two node_ids, without actually being nodes.

[the remark this answers is not in the document]
Comment (wildarch) [reply oak_runtime/src/lib.rs]: Aah, sorry you are right, I think the `AuxServer` does not need to set this.

I will remove it, thanks 😄

File: examples/abitest/client/cpp/abitest.cc [hunk 3]
```diff
@@ -195,25 +195,25 @@ bool run_aux_tests() {
   success &= check_page(&introspect, "/graph");
   success &= check_page(&introspect, "/objcount");
   success &= check_page(&introspect, "/node/0");
-  success &= check_page(&introspect, "/node/1");
+  success &= check_page(&introspect, "/node/3");
```
Comment (rbehjati) [inline examples/abitest/client/cpp/abitest.cc:198]: Is this change still needed? Or does it now work with "/node/1".

File: oak_runtime/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rbehjati) [reply oak_runtime/src/lib.rs]: Thanks :)

File: examples/abitest/client/cpp/abitest.cc (hunk 3 above)
Comment (wildarch) [reply examples/abitest/client/cpp/abitest.cc]: Aah, looks like it works with 1 now. I'm not sure what changed, do you have any ideas?

Comment (rbehjati) [reply examples/abitest/client/cpp/abitest.cc]: I think that call to `proxy_for_new_node` in `AuxServer` was unnecessarily increasing the `node_id`.

**Review by rbehjati — APPROVED** [review]
LGTM if the tests pass :)

File: examples/aggregator/client/android/cpp/client.cc [hunk 4]
```diff
@@ -55,7 +55,7 @@ JNIEXPORT void JNICALL Java_com_google_oak_aggregator_MainActivity_createChannel
   // The particular value corresponds to the hash on the `aggregator.wasm` line in
   // https://github.com/project-oak/oak/blob/hashes/reproducibility_index.
   oak::label::Label label = oak::WebAssemblyModuleHashLabel(
-      absl::HexStringToBytes("a3d55c1e3af8dc34dfe5a3736bfcbd7db8541a607a33089730e81eda3d5b63bc"));
+      absl::HexStringToBytes("4b652112bb9904976fd5a215a4df081956165f1dbc19bf6c77e0eef05c4bc256"));
```
Comment (rbehjati) [inline examples/aggregator/client/android/cpp/client.cc]: Out of curiosity: do you remember why you had to make this change (or the other change that is reverted in this commit), and where the other hash came from? We had a strange discussion earlier involving exactly these two hash values: https://project-oak.slack.com/archives/CHE9E13C3/p1615996966001400?thread_ts=1615996957.001300&cid=CHE9E13C3
